### PR TITLE
Enable bool and int32 tests | test(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
Enable all bool and int32 tests for torchlib functions and mark tests with xfail as appropriate.

Fix

- Fix `aten::equal` to support BOOL input.
- Simplify `aten::any`, `aten::all`
- Mark `erf*` as floating point only.
- Mark xfail on `aten::sum` and `aten::cumsum` which use int64 as the accumulator type for int32 inputs.
- Turn all `keepdims=0` -> `keepdims=False`

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -2370,6 +2391,18 @@ def aten_equal(self: TTensor, other: TTensor) -> BOOL:
     return op.Equal(sum_of_abs, 0)
 
 
+@torch_op("aten::equal")
+def aten_equal_bool(self: BOOL, other: BOOL) -> BOOL:
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:2395]: why can't we directly onnx Equal self and other? Please also comment it thx.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:2395]: I don't know what I was thinking. For some reason I thought equal didn't work on bools. Updated.

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 2]
```diff
@@ -1792,6 +1834,7 @@ def _where_input_wrangler(
 ops_test_common.duplicate_opinfo(OPS_DB, "atleast_1d", ("atleast_1d_single_tensor",))
 ops_test_common.duplicate_opinfo(OPS_DB, "atleast_2d", ("atleast_2d_single_tensor",))
 ops_test_common.duplicate_opinfo(OPS_DB, "atleast_3d", ("atleast_3d_single_tensor",))
+ops_test_common.duplicate_opinfo(OPS_DB, "equal", ("equal_bool",))
 ops_test_common.duplicate_opinfo(OPS_DB, "cat", ("concat", "concatenate"))
 ops_test_common.duplicate_opinfo(OPS_DB, "clone", ("lift_fresh_copy",))
 ops_test_common.duplicate_opinfo(OPS_DB, "full_like", ("full_like_dtype",))
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1840]:
```suggestion
ops_test_common.duplicate_opinfo(OPS_DB, "cat", ("concat", "concatenate"))
ops_test_common.duplicate_opinfo(OPS_DB, "clone", ("lift_fresh_copy",))
ops_test_common.duplicate_opinfo(OPS_DB, "equal", ("equal_bool",))
ops_test_common.duplicate_opinfo(OPS_DB, "full_like", ("full_like_dtype",))
```

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/core.py:2395]: Probably because how aten_equal was implemented up there? It is also not directly onnx equal. This question applies there as well. I assume there is something I'm missing.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:2395]: Hmm what's going on. Let me try to merge that.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Done. Thanks for catching this!

Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Okay, I realized the torch.equal is probably returning single True/False for the result, while ONNX one does elementwise comparison. Let's add the comment ..

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Added

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -2364,10 +2380,13 @@ def aten_eq(self: TTensor, other: TTensor) -> BOOL:
 def aten_equal(self: TTensor, other: TTensor) -> BOOL:
     """equal(Tensor self, Tensor other) -> bool"""
 
-    sub_self_other = op.Sub(self, other)
-    abs_sub = op.Abs(sub_self_other)
-    sum_of_abs = op.ReduceSum(abs_sub, keepdims=0)
-    return op.Equal(sum_of_abs, 0)
+    # NOTE: Torch aten::equal returns a single Boolean while ONNX Equal is elementwise.
+    # The equivalent Torch op with ONNX Equal is aten::eq.
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:2384]: nice

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 4]
```diff
@@ -2364,10 +2380,13 @@ def aten_eq(self: TTensor, other: TTensor) -> BOOL:
...
+    elementwise_equal = op.Equal(self, other)
+    elementwise_equal_int = op.Cast(elementwise_equal, to=INT64.dtype)
+    # ReduceMax does not support bool. So we cast to int64
+    all_equal = op.ReduceMin(elementwise_equal_int, keepdims=0)
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:2388]: nit: I saw in other place `keepdims=0` was changed to `keepdims=False`. Is there a preference between the two expression?

**Review by BowenBao — APPROVED** [review]

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 5]
```diff
@@ -1247,22 +1274,29 @@ def _where_input_wrangler(
...
+    ),
     TorchLibOpInfo("xlogy", special_ops.aten_special_xlogy),
     TorchLibOpInfo("zeros", core_ops.aten_zeros),
     TorchLibOpInfo(
         "arange_start_step",
         core_ops.aten_arange_start_step,
         trace_only=True,
-    ).xfail(
+    )
+    .xfail(dtypes=(torch.int32,), reason="fixme: output shape mismatch in edge cases.")
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1288]: rounding problem?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 6]
```diff
@@ -338,41 +358,36 @@ def aten_angle(self: TensorType) -> TensorType:
...
-        # cannot cast to INT64 because 0.1 will be cast to 0, then convert to false
         self_bool = op.Cast(self, to=BOOL.dtype)
-        # op.ReduceMax() in next step cannot calculate BOOL value, so convert to INT64
+        # op.ReduceMax() in the next step cannot process BOOL inputs, so convert to INT64
         self_int = op.Cast(self_bool, to=INT64.dtype)
         # Change dim from int to INT64[1]
         dims = op.Reshape(dim, op.Constant(value_ints=[-1]))
-        result_max = op.ReduceMax(self_int, dims, keepdims=keepdim, noop_with_empty_axes=0)
-        result = op.Greater(result_max, op.Constant(value_int=0))
+        any_true = op.ReduceMax(self_int, dims, keepdims=keepdim)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/core.py:389]: As an aside (not for this PR): I think we should introduce an equivalent op in ONNX: either simply extend ReduceMax and ReduceMin for boolean types (treating False < True) or introduce new boolean ops for ReduceAnd, ReduceOr. I think the first is simpler.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:389]: This would be very helpful

> I think the first is simpler.

I think so too!

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1288]: Potentially - it was an off-by-one error

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:2388]: I personally prefer False. Let me change elsewhere

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Done

**Comment by justinchuby** [comment]
Requires https://github.com/microsoft/onnxscript/pull/937

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:389]: https://github.com/onnx/onnx/issues/5452
